design-proposals/emts-scale-provisioning.md

```diff
@@ -0,0 +1,321 @@
+2. ENs on site start PXE boot. They initiate DHCP discovery.
+3. DHCP requests are intercepted by Tinkerbell SMEE. It replies with a TFTP address pointing to itself. The Tinkerbell SMEE provides default iPXE environment (e.g., `ipxe.efi`).
+   It supports a range of different PXE architectures (see [link](https://github.com/tinkerbell/smee/blob/c0443b87cec44bf5d7d0028c876ea3be4f4d3d47/internal/dhcp/dhcp.go#L44)).
+4. `ipxe.efi` is downloaded to PXE environment.
+5. EN leaves PXE context and loads iPXE execution environment.
+6. iPXE must perform a separate DHCP discovery as EN left PXE context. It sends DHCP requests with another User-Class (`Tinkerbell`).
+7. This time, Tinkerbell SMEE replies with the HTTP URL to `boot.ipxe`. The decision on what to return is based on the User-Class.
+8. iPXE gets DHCP reply with HTTP URL to EIM's Provisioning Nginx. Note that this is a plain HTTP endpoint (without TLS) because the `ipxe.efi` provided by SMEE
+   doesn't include orchestrator's CA certificate and the download would fail. We elaborate more on this below. 
+9. iPXE downloads `boot.ipxe` via HTTP from Provisioning Nginx. `boot.ipxe` is our EIM-provided iPXE script that downloads and boots into Micro-OS.
+10. `boot.ipxe` is downloaded.
+11. iPXE chain-loads to `boot.ipxe`.
+12. `boot.ipxe` downloads Micro-OS image from the HTTP endpoint of Provisioning Nginx.
+
+Once Micro-OS is downloaded, EN boots into it and the standard process is continued - device discovery is done and Tinkerbell workflow started to provision target OS.
+
+The following modifications to EIM must be made to support the above workflow:
+
+1. **Enable Tinkerbell SMEE with proper configuration**
+
+By default, Tinkerbell SMEE relies on MAC addresses to uniquely identify PXE-booting machines and customize the iPXE script per machine.
+In the case of EIM, we use a static iPXE script that is not customized per Edge Node. Therefore, we can avoid using MAC address as a unique identifier
+and it will let use avoid adding MAC address as another EN identifier during pre-registration. We also need to configure Tinkerbell to provide URL to `boot.ipxe` via DHCP reply.
+
+The following modifications are needed:
+
+- Tinkerbell SMEE must be enabled via infra-charts to provide DHCP and TFTP server
+- Tinkerbell SMEE must be configured with the following flags to avoid lookups by MAC address and to use our EIM iPXE script:
+  - `-dhcp-mode=auto-proxy`
+  - `-dhcp-http-ipxe-script-url=<URL to iPXE script on Provisioning NGINX>` (e.g., `http://tinkerbell-nginx.CLUSTER_DOMAIN/tink-stack/boot.ipxe`)
+  - `-dhcp-http-ipxe-script-prepend-mac=false`
+
+2. **Expose Provisioning Nginx via HTTP for on-prem EIM standalone only**
+
+Currently, Provisioning Nginx is deployed behind HTTPS and the iPXE binary must be built with orchestrator's CA certificate for successful handshake.
+Tinkerbell SMEE obviously doesn't include the EMF's CA certificate. Therefore, any communication over HTTPS is not feasible.
```

Comment:
Good to check with SAFE team once to go ahead with HTTP provided local network is protected with firewall.

Reply:
After some experiments, the decision is to use off-the-shelf DHCP server to provide "our" `signed_ipxe.efi`, see PoC charts: https://github.com/open-edge-platform/infra-charts/pull/53